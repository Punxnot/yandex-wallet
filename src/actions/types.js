export const USER_LOGIN_SUCCESS = 'USER_LOGIN_SUCCESS';
export const USER_LOGIN_FAILURE = 'USER_LOGIN_FAILURE';
export const USER_LOGOUT = 'USER_LOGOUT';

export const CARDS_FETCH_STARTED = 'CARDS_FETCH_STARTED';
export const CARDS_FETCH_SUCCESS = 'CARDS_FETCH_SUCCESS';
export const CARDS_FETCH_FAILED = 'CARDS_FETCH_FAILED';

export const CARD_DELETE_STARTED = 'CARD_DELETE_STARTED';
export const CARD_DELETE_SUCCESS = 'CARD_DELETE_SUCCESS';
export const CARD_DELETE_FAILED = 'CARD_DELETE_FAILED';

export const CARD_FETCH_SUCCESS = 'CARD_FETCH_SUCCESS';
export const CARD_FETCH_FAILED = 'CARD_FETCH_FAILED';

export const TRANS_FETCH_STARTED = 'TRANS_FETCH_STARTED';
export const TRANS_FETCH_FAILED = 'TRANS_FETCH_FAILED';
export const TRANS_FETCH_SUCCESS = 'TRANS_FETCH_SUCCESS';

export const ACTIVE_CARD_CHANGED = 'ACTIVE_CARD_CHANGED';

export const MOBILE_PAY_SUCCESS = 'MOBILE_PAY_SUCCESS';
export const MOBILE_PAY_FAILED = 'MOBILE_PAY_FAILED';
export const MOBILE_PAY_REPEAT = 'MOBILE_PAY_REPEAT';

export const PREPAID_PAY_SUCCESS = 'PREPAID_PAY_SUCCESS';
export const PREPAID_PAY_FAILED = 'PREPAID_PAY_FAILED';
export const PREPAID_PAY_REPEAT = 'PREPAID_PAY_REPEAT';

export const WITHDRAW_PAY_SUCCESS = 'WITHDRAW_PAY_SUCCESS';
export const WITHDRAW_PAY_FAILED = 'WITHDRAW_PAY_FAILED';
export const WITHDRAW_PAY_REPEAT = 'WITHDRAW_PAY_REPEAT';
