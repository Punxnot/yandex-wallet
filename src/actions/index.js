import { push } from 'react-router-redux';

import * as action from './types';

import axios from 'axios';

const ROOT_URL = '/api';
// const LOGIN_URL = `${ROOT_URL}/signin`;
// const SIGNUP_URL = `${ROOT_URL}/signup`;

export const signingUser = () => {
    return dispatch => {
        dispatch({
            type: action.USER_LOGIN_SUCCESS
        });

        dispatch(push('/cards'));
    }
};

export const signOutUser = () => {
    return dispatch => {
        dispatch({
            type: action.USER_LOGOUT
        });

        dispatch(push('/'));
    }
};

/**
* Вытаскивает данные по картам пользователя
* 
* @returns 
*/
export const fetchCards = () => {
    return dispatch => {
        axios
            .get(`${ROOT_URL}/cards`, {
                headers: {
                    authorization: 'JWT ' + localStorage.getItem('token')
                }
            })
            .then(res => {
                dispatch({
                    type: action.FETCH_CARDS_SUCCESS,
                    payload: res.data
                });
                if (res.data[0].id)
                    dispatch(changeActiveCard(res.data[0].id));
            })
            .catch(res => {
                dispatch({
                    type: action.FETCH_CARDS_FAILED,
                    payload: res.response.data.message ? res.response.data.message : res.response.data
                });
                console.log(res.response.data.message ? res.response.data.message : res.response.data);
            });
    };
};

/**
* Вытаскивает обновленные данные по карте пользователя
* 
* @param {Number} id 
* @returns 
*/
export const fetchCard = (id) => {
    return dispatch => {
        axios
            .get(`${ROOT_URL}/cards/${id}`, {
                headers: {
                    authorization: 'JWT ' + localStorage.getItem('token')
                }
            })
            .then(res => {
                dispatch({
                    type: action.FETCH_CARD_SUCCESS,
                    payload: res.data
                });
                dispatch(fetchTransactions(id));
            })
            .catch(res => {
                dispatch({
                    type: action.FETCH_CARD_FAILED,
                    payload: res.response.data.message ? res.response.data.message : res.response.data
                });
                console.log(res.response.data.message ? res.response.data.message : res.response.data);
            });
    };
};

/**
* Вытаскивает данные по картам пользователя
* 
* @returns 
*/
export const fetchTransactions = id => {
    return dispatch => {
        axios
            .get(`${ROOT_URL}/cards/${id}/transactions`, {
                headers: {
                    authorization: 'JWT ' + localStorage.getItem('token')
                }
            })
            .then(res => {
                dispatch({
                    type: action.FETCH_TRANS_SUCCESS,
                    payload: res.data
                });
            })
            .catch(res => {
                dispatch({
                    type: action.FETCH_TRANS_FAILED,
                    payload: res.response.data.message ? res.response.data.message : res.response.data
                });
                console.log(res.response.data.message ? res.response.data.message : res.response.data);
            });
    };
};

export const changeActiveCard = id => {
    return dispatch => {
        dispatch({
            type: action.CHANGE_ACTIVE_CARD,
            payload: id
        });
        dispatch(fetchTransactions(id));
    }
};

export const mobilePayment = (transaction, id) => {
    // формируем транзакцию
    const apiTransaction = {
        type: 'paymentMobile',
        data: transaction.phoneNumber,
        time: Date.now() / 1000,
        sum: transaction.sum < 0 ? transaction.sum : transaction.sum * -1
    };

    return dispatch => {
        axios
            .post(`${ROOT_URL}/cards/${id}/transactions`, apiTransaction, {
                headers: {
                    authorization: 'JWT ' + localStorage.getItem('token')
                }
            })
            .then(res => {
                if (res.status === 201) {
                    dispatch({
                        type: action.MOBILE_PAY_SUCCESS,
                        payload: transaction
                    });
                    dispatch(fetchCard(id));
                }
            })
            .catch(res => {
                dispatch({
                    type: action.MOBILE_PAY_FAILED,
                    payload: {
                        error: res.response.data.message ? res.response.data.message : res.response.data,
                        transaction
                    }
                });
                console.log(res.response.data.message ? res.response.data.message : res.response.data);
            });
    }
};

export const repeateMobileTransfer = () => {
    return dispatch => {
        dispatch({
            type: action.MOBILE_PAY_WISH_TO_REPEAT
        });
    }
};
