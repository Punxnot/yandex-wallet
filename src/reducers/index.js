import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';

import authReducer from './auth_reducer';
import cardsReducer from './cards_reducer';
import mrs_reducer from './mrs_reducer';
import transactionReducer from './transactions_reducer';
import mobilePaymentReducer from './mobile_payment_reducer';
import prepaidPaymentReducer from './prepaid_payment_reducer';
import withdrawPaymentReducer from './withdraw_payment_reducer';

const rootReducer = combineReducers({
    auth: authReducer,
    routing: routerReducer,
    cards: cardsReducer,
    transactions: transactionReducer,
    payments: combineReducers({
        mobilePayment: mobilePaymentReducer,
        prepaidPayment: prepaidPaymentReducer,
        withdrawPayment: withdrawPaymentReducer
    }),
    mrs: mrs_reducer
});

export default rootReducer;
