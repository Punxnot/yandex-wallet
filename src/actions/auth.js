import { push } from 'react-router-redux';
import axios from 'axios';

import { fetchCards } from './cards';
import * as action from './types';

const ROOT_URL = '/api/auth';
const VERIFY_URL = `${ROOT_URL}/verify`;
const SIGNIN_URL = `${ROOT_URL}/signin`;
const SIGNUP_URL = `${ROOT_URL}/signup`;

/**
 * Разлогинивается, если заметит изменение токена в другом окне
 * 
 */
const setStorageEvents = dispatch => {
    if (window) window.addEventListener('storage', e => {
            if (e.newValue === null && e.key === 'token')
                dispatch(signOutUser());
        });
}

export const verifyToken = () => {
    return async dispatch => {
        const token = localStorage.getItem('token');
        if (token) {
            setStorageEvents(dispatch);

            dispatch({
                type: action.USER_TOKEN_VERIFY_START
            });

            try {
                const response = await axios
                    .get(VERIFY_URL, {
                        headers: {
                            authorization: 'JWT ' + token
                        }
                    });

                dispatch({
                    type: action.USER_LOGIN_SUCCESS,
                    payload: response.data.user
                });

                dispatch(fetchCards());

                dispatch(push('/'));
            } catch (err) {
                let message = 'Сеанс закончился';
                if (err.response.status === 401)
                    if (err.response.data.message === 'invalid token')
                        message = 'Ошибка авторизации';

                dispatch(signOutUser(message));
            }
        }
    }
}

export const signInUser = ({email, password}) => {
    return async dispatch => {
        try {
            const response = await axios
                .post(SIGNIN_URL, {
                    email,
                    password
                });

            localStorage.setItem('token', response.data.token);

            dispatch({
                type: action.USER_LOGIN_SUCCESS,
                payload: response.data.user
            });

            dispatch(fetchCards());

            dispatch(push('/'));
        } catch (response) {
            dispatch({
                type: action.USER_LOGIN_FAILURE,
                payload: response.response.data.message ? response.response.data.message : response.response.data
            });
            console.log(response.response.data.message ? response.response.data.message : response.response.data);
        }
    }
};

export const signOutUser = err => {
    return dispatch => {
        localStorage.removeItem('token');

        dispatch({
            type: action.USER_LOGOUT,
            payload: err
        });

        dispatch(push('/signin'));
    }
};
