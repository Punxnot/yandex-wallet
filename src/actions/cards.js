import * as action from './types';
import { signOutUser } from './auth';
import { fetchTransactions } from './transactions';

import axios from 'axios';

const ROOT_URL = '/api';

/**
* Добавляет новую карту пользователю
* 
*/
export const addCard = (cardNumber, currency, exp, name) => {
    return async dispatch => {
        try {
            dispatch({
                type: action.CARD_ADD_STARTED
            });

            const data = {
                cardNumber,
                currency,
                exp,
                name
            };

            const response = await axios
                .post(`${ROOT_URL}/cards`, data, {
                    headers: {
                        authorization: 'JWT ' + localStorage.getItem('token')
                    }
                });

            if (response.status === 201) {
                dispatch({
                    type: action.CARD_ADD_SUCCESS
                });
                dispatch(fetchCards());
            }
            else dispatch({
                    type: action.CARD_ADD_FAILED,
                    payload: 'Что то пошло не так...'
                });
        } catch (err) {
            if (err.response.status === 401)
                dispatch(signOutUser('Ошибка авторизации'));

            dispatch({
                type: action.CARD_ADD_FAILED,
                payload: err.response.data.message ? err.response.data.message : err.response.data
            });
            console.log(err.response.data.message ? err.response.data.message : err.response.data);
        }
    }
}

/**
* Вытаскивает данные по картам пользователя
* 
*/
export const fetchCards = () => {
    return async dispatch => {
        try {
            dispatch({
                type: action.CARDS_FETCH_STARTED
            });

            const response = await axios
                .get(`${ROOT_URL}/cards`, {
                    headers: {
                        authorization: 'JWT ' + localStorage.getItem('token')
                    }
                });

            dispatch({
                type: action.CARDS_FETCH_SUCCESS,
                payload: response.data
            });

            if (response.data.length > 0)
                if (response.data[0].id)
                    dispatch(changeActiveCard(response.data[0].id));

        } catch (err) {
            if (err.response.status === 401)
                dispatch(signOutUser('Ошибка авторизации'));

            dispatch({
                type: action.CARDS_FETCH_FAILED,
                payload: err.response.data.message ? err.response.data.message : err.response.data
            });
            console.log(err.response.data.message ? err.response.data.message : err.response.data);
        }
    }
}

/**
* Вытаскивает данные по карте пользователя
* 
* @param {Number} id 
*/
export const fetchCard = id => {
    return async dispatch => {
        try {
            const response = await axios
                .get(`${ROOT_URL}/cards/${id}`, {
                    headers: {
                        authorization: 'JWT ' + localStorage.getItem('token')
                    }
                });

            dispatch({
                type: action.CARD_FETCH_SUCCESS,
                payload: response.data
            });
        } catch (err) {
            if (err.response.status === 401)
                dispatch(signOutUser('Ошибка авторизации'));

            dispatch({
                type: action.CARD_FETCH_FAILED,
                payload: err.response.data.message ? err.response.data.message : err.response.data
            });
            console.log(err.response.data.message ? err.response.data.message : err.response.data);
        }
    }
}

/**
* Удаляет карту пользователя
* 
* @param {Number} id 
* @returns 
*/
export const deleteCard = id => {
    return async dispatch => {
        try {
            dispatch({
                type: action.CARD_DELETE_STARTED
            });

            const response = await axios
                .delete(`${ROOT_URL}/cards/${id}`, {
                    headers: {
                        authorization: 'JWT ' + localStorage.getItem('token')
                    }
                });

            if (response.status === 200) {
                dispatch({
                    type: action.CARD_DELETE_SUCCESS
                });
                dispatch(fetchCards());
            }
            else dispatch({
                    type: action.CARD_DELETE_FAILED,
                    payload: 'Что то пошло не так...'
                });
        } catch (err) {
            if (err.response.status === 401)
                dispatch(signOutUser('Ошибка авторизации'));

            dispatch({
                type: action.CARD_DELETE_FAILED,
                payload: err.response.data.message ? err.response.data.message : err.response.data
            });
            console.log(err.response.data.message ? err.response.data.message : err.response.data);
        }
    }
}

export const changeActiveCard = id => (dispatch, getState) => {
    const {activeCardId} = getState().cards;

    if (activeCardId === id) return;

    dispatch({
        type: action.ACTIVE_CARD_CHANGED,
        payload: id
    });
    dispatch(fetchTransactions(id));
}
