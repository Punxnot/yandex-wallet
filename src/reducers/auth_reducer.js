import * as actions from '../actions/types';

export const authInitialState = {
    isAuth: false,
    userName: null,
    error: null
};

const authReducer = (state = authInitialState, {type, payload}) => {
    switch (type) {
        case actions.USER_LOGIN_SUCCESS:
            return {
                ...state,
                isAuth: true,
                error: null,
                userName: payload
            };
        case actions.USER_LOGIN_FAILURE:
            return {
                ...state,
                isAuth: false,
                error: payload
            };
        case actions.USER_LOGOUT:
            return authInitialState

        default:
            return state;
    }
};

export default authReducer;
