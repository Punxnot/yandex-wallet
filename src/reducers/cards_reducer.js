import * as actions from '../actions/types';

/**
 * Возвращает массив с обновленным элементом
 * 
 * @param {Array} array 
 * @param {Object} action 
 * @returns {Array}
 */
const addOrUpdateObjectInArray = (array, newItem) => {
    let isNew = true;
    const newArray = array.map((item) => {
        if (item.id !== newItem.id) return item;
        isNew = false;
        return newItem;
    });
    if (isNew) newArray.push(newItem);
    return newArray;
};

export const cardsInitialState = {
    data: [],
    error: null,
    isLoading: false,
    activeCardId: null
};

const cardsReducer = (state = cardsInitialState, {type, payload}) => {
    switch (type) {
        case actions.CARDS_FETCH_STARTED:
            return {
                ...state,
                isLoading: state.data.length === 0 ? true : false
            }

        case actions.CARD_ADD_FAILED:
            return {
                ...state,
                error: payload
            }

        case actions.CARDS_FETCH_SUCCESS:
            return {
                ...state,
                data: payload,
                error: null,
                isLoading: false
            }

        case actions.CARDS_FETCH_FAILED:
            return {
                ...state,
                error: payload,
                isLoading: false
            }

        case actions.USER_LOGOUT:
            return cardsInitialState

        case actions.ACTIVE_CARD_CHANGED:
            return {
                ...state,
                activeCardId: payload
            }

        case actions.CARD_FETCH_SUCCESS:
            return {
                ...state,
                data: addOrUpdateObjectInArray(state.data, payload)
            }

        case actions.CARD_FETCH_FAILED:
            return {
                ...state,
                error: payload
            }

        default:
            return state
    }
}

export default cardsReducer;
