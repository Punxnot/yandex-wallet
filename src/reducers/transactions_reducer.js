import * as actions from '../actions/types';

export const transactionsInitialState = {
    data: [],
    error: null,
    isLoading: true
};

const transactionsReducer = (state = transactionsInitialState, {type, payload}) => {
    switch (type) {
        case actions.TRANS_FETCH_STARTED:
            return {
                ...state,
                isLoading: state.data.length === 0 ? true : false
            }

        case actions.TRANS_FETCH_SUCCESS:
            return {
                ...state,
                data: payload,
                error: null,
                isLoading: false
            }

        case actions.TRANS_FETCH_FAILED:
            return {
                ...state,
                error: payload,
                isLoading: false
            }

        case actions.CARDS_FETCH_FAILED:
            return transactionsInitialState

        case actions.USER_LOGOUT:
            return transactionsInitialState

        default:
            return state
    }
};

export default transactionsReducer;
