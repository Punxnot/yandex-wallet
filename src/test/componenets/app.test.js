import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import renderer from 'react-test-renderer'

import App from '../../components/app';
import { cardsInitialState } from '../../reducers/cards_reducer';
import { authInitialState } from '../../reducers/auth_reducer';
import { transactionsInitialState } from '../../reducers/transactions_reducer';
// import * as cardActions from '../../actions/cards';
// import * as transactionsActions from '../../actions/transactions';
import * as types from '../../actions/types';

import { API_END_POINT, cards, transactions, initLocalStorage, testAsyncAction } from '../common';

const {id} = cards[0];

beforeAll(() => {
  configure({
    adapter: new Adapter()
  });

  initLocalStorage();

  const mock = new MockAdapter(axios);

  mock.onGet(API_END_POINT + 'cards').reply(200, cards);
  mock.onDelete(API_END_POINT + `cards/${id}`).reply(200);
  mock.onGet(API_END_POINT + `cards/${id}/transactions`).reply(200, transactions);
});

describe('Render whole App with initital state', () => {
  let store,
    wrapper;

  beforeEach(() => {
    const initialState = {
      auth: authInitialState,
      cards: cardsInitialState,
      transactions: transactionsInitialState
    };

    const mockStore = configureStore([thunk]);

    store = mockStore(initialState);

    wrapper = mount(<Provider store={ store }>
                      <App />
                    </Provider>);
  });

  it('+++ capturing Snapshot of App', () => {
    const renderedValue = renderer.create(<Provider store={ store }>
                                            <App />
                                          </Provider>).toJSON()
    expect(renderedValue).toMatchSnapshot();
  });

  it('+++ render the connected component', () => {
    expect(wrapper.find(App).length).toEqual(1);
  });

  it('+++ check initial action on start', done => {

    const expectedActions = [
      {
        type: types.FETCH_CARDS
      },
      {
        type: types.FETCH_CARDS_SUCCESS,
        payload: cards
      },
      {
        type: types.ACTIVE_CARD_CHANGE,
        payload: id
      },
      {
        type: types.FETCH_TRANS
      }
    ];

    store.dispatch(testAsyncAction())
      .then(() => {
        const actions = store.getActions();
        expect(actions).toEqual(expectedActions);
        done();
      });
  });
});
