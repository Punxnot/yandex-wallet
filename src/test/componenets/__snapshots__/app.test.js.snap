// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render whole App with initital state +++ capturing Snapshot of App 1`] = `
<div
  className="css-Wallet-34j4y90 css-46dg62"
>
  <div
    className="css-Layout-n62tmw0 css-oqhttl"
  >
    <div
      className="css-Logo-n62tmw1 css-g3dddy"
    />
    <div
      className="css-Edit-n62tmw2 css-1rly8q8"
      onClick={[Function]}
    />
    <div
      className="css-CardsList-n62tmw3 css-14qlmqd"
    >
      <div
        className="css-CardLayout-1ezavq10 css-NewCardLayout-1ezavq14 css-9okkk3"
      />
    </div>
    <footer
      className="css-Footer-n62tmw4 css-1s8mtv6"
    >
      Yamoney Node School
    </footer>
  </div>
  <div
    className="css-CardPane-34j4y91 css-i5j6vn"
  >
    <header
      className="css-HeaderLayout-1f0bsvn0 css-6irvbk"
    >
      <div
        className="css-User-1so6ulr1 css-1l3dfk3"
      >
        <img
          className="css-Avatar-1so6ulr2 css-1xfhnhw"
          src="/assets/avatar.png"
        />
        Samuel Johnson
        <div
          className="css-DropDown-1so6ulr0 css-1qvsddu"
        >
          <a
            onClick={[Function]}
          >
            Выйти
          </a>
        </div>
      </div>
    </header>
  </div>
</div>
`;
